[PHPMD] Add phpmd rule with ignoring of unused required parameters in the plugins

Den4ik opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->
This PR improve phpmd `UnusedFormalParameter` rule with ignoring of required unused parameters in the plugins.

- `around` and `after` plugins has 2 required params `$subject` and `$proceed` (or `$result` for `after` plugin).
- `before` plugin has 1 required param `$subject`.

This PR override standard `UnusedFormalParameter` rule and allow don't use `@SuppressWarnings(PHPMD.UnusedFormalParameter)` in the plugins.

### Related Pull Requests
<!-- related pull request placeholder -->

### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->

### Manual testing scenarios (*)
<!---
    Please provide a set of unambiguous steps to test the proposed code change.
    Giving us manual testing scenarios will help with the processing and validation process.
-->
1. Configure phpmd
2. Remove `@SuppressWarnings(PHPMD.UnusedFormalParameter)` from some plugin
3. Add `PluginInterface` implementation into the class
4. Run phpmd. Expected no `UnusedFormalParameter` errors

### Questions or comments
<!---
	If relevant, here you can ask questions or provide comments on your pull request for the reviewer
	For example if you need assistance with writing tests or would like some feedback on one of your development ideas
-->

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)


### Resolved issues:
1. [x] resolves magento/magento2#37914: [PHPMD] Add phpmd rule with ignoring of unused required parameters in the plugins

Den4ik committed:
[PHPMD] Add phpmd rule with ignoring of unused required parameters in the plugins

Signed-off-by: Denis Kopylov <dkopylov@magenius.team>

Den4ik committed:
Add copyright

Signed-off-by: Denis Kopylov <dkopylov@magenius.team>

ihor-sviziev commented on dev/tests/static/framework/Magento/CodeMessDetector/Rule/UnusedCode/UnusedFormalParameter.php:
```diff
@@ -0,0 +1,106 @@
+            /** @var ClassNode $classNode */
+            $classNode = $node->getParentType();
+            $implementInterfaces = class_implements($classNode->getFullQualifiedName());
+            if (array_key_exists(PluginInterface::class, $implementInterfaces)) {
```
I think introducing a special interface to fix the static test failures is the wrong way to do it. Can we just check if the class is located in the `*\Plugin\*` namespace?

ihor-sviziev replied:
Could you also cover your change with unit test? 
Example you can see here:
https://github.com/phpmd/phpmd/tree/master/src/test/php/PHPMD/Rule/CleanCode

Den4ik replied:
> I think introducing a special interface to fix the static test failures is the wrong way to do it. Can we just check if the class is located in the *\Plugin\* namespace?

I thought about it, but in fact we could have non plugin classes in `Plugin` namespace.

Den4ik replied:
> Could you also cover your change with unit test?

Yep if idea will be approved by core team

bgorski replied:
Honestly, while I totally agree with the idea of phpmd skipping those parameters in the check, I also think that introducing a new marker interface for this is a wrong approach. 

While I don't have any experience in writing or improving phpmd checks myself, I believe that you should be able to parse di.xml files from the same module to determine if a class is a plugin or not. Am I correct with this assumption?

Effort-wise, with a marker interface, you just trade the necessity of writing `SuppressWarnings(PHPMD.UnusedFormalParameter)` for the necessity of writing `implements \Magento\Framework\Interception\PluginInterface`.

Den4ik replied:
@bgorski I thought about parsing of di.xml but it's very expensive only for tests

Den4ik replied:
I'll try to implement parsing and check preformance

ihor-sviziev replied:
I think checking the namespace will be a good enough solution. The edge case, when the plugin not in the `Plugin` namespace still could be ignored

Den4ik committed:
Detect plugin in namespace

Signed-off-by: Denis Kopylov <dkopylov@magenius.team>

ihor-sviziev commented on dev/tests/static/framework/Magento/CodeMessDetector/Rule/UnusedCode/UnusedFormalParameter.php:
```diff
@@ -0,0 +1,115 @@
+
+namespace Magento\CodeMessDetector\Rule\UnusedCode;
+
+use Magento\Framework\Interception\PluginInterface;
```
Looks like this class left from previous iteration and doesn't exist 
```suggestion
```

Den4ik committed:
Remove interface declaration

Signed-off-by: Denis Kopylov <dkopylov@magenius.team>

ihor-sviziev commented on dev/tests/static/framework/Magento/CodeMessDetector/Rule/UnusedCode/UnusedFormalParameter.php:
```diff
@@ -0,0 +1,114 @@
+     */
+    private function removeVariablesUsedInPlugins(AbstractNode $node)
+    {
+        if ($node instanceof MethodNode) {
```
I think there it would be better to revert checks and add `return`.

ihor-sviziev commented on dev/tests/static/framework/Magento/CodeMessDetector/Rule/UnusedCode/UnusedFormalParameter.php:
```diff
@@ -0,0 +1,114 @@
+        if ($node instanceof MethodNode) {
+            /** @var ClassNode $classNode */
+            $classNode = $node->getParentType();
+            if ($this->isPluginClass($classNode->getNamespaceName())) {
```
same here

ihor-sviziev reviewed (changes requested):
Please cover changes with some unit test

Den4ik committed:
Implement suggested changes

Signed-off-by: Denis Kopylov <dkopylov@magenius.team>

Den4ik commented:
@ihor-sviziev I'll work on tests in a week

Den4ik committed:
Refactoring + added unit tests

Signed-off-by: Denis Kopylov <dkopylov@magenius.team>

Den4ik commented:
@magento run all tests

Den4ik commented:
@ihor-sviziev @bgorski I updated PR with some refactoring and added unit tests. Please approve when all tests will be green and you don't have any suggestions.

Den4ik committed:
Fix static tests

Signed-off-by: Denis Kopylov <dkopylov@magenius.team>

Den4ik commented:
@magento run all tests

Den4ik commented:
@magento run Functional Tests EE, Functional Tests CE

Den4ik commented:
@magento run Functinal Tests B2B

ihor-sviziev commented:
@magento run Functinal Tests B2B

Den4ik commented:
@magento run Functional Tests B2B

ihor-sviziev commented on dev/tests/static/framework/Magento/CodeMessDetector/Test/Unit/Rule/UnusedCode/UnusedFormalParameterTest.php:
```diff
@@ -0,0 +1,182 @@
+            ]
+        );
+
+        /*$methodNode->expects($this->once())
```
I think it could be removed

Den4ik replied:
Oh, yep

ihor-sviziev commented on dev/tests/static/framework/Magento/CodeMessDetector/Rule/UnusedCode/UnusedFormalParameter.php:
```diff
@@ -0,0 +1,101 @@
+     */
+    private function isFunctionNameStartingWith(MethodNode $node, string $name): bool
+    {
+        return (0 === strpos($node->getImage(), $name));
```
Interesting... why it's named as getImage? 😃

Den4ik replied:
Both `getName` and `getImage` contain `return $this->name`.
Tests in phpmd source use `getImage`

ihor-sviziev replied:
Could you add this as a comment?

Den4ik committed:
Remove nused code

Signed-off-by: Denis Kopylov <dkopylov@magenius.team>

Den4ik commented:
@magento run all tests

Den4ik committed:
Update doc block

Signed-off-by: Denis Kopylov <dkopylov@magenius.team>

Den4ik commented:
@magento run all tests

Den4ik committed:
Update doc block

Signed-off-by: Denis Kopylov <dkopylov@magenius.team>

Den4ik commented:
@magento run all tests

Den4ik commented:
@magento run Functional Tests EE, Functional Tests B2B

Den4ik commented:
@magento run Functional Tests B2B

magento-engcom-team commented:
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-9223](https://jira.corp.magento.com/browse/ENGCOM-9223) has been created to process this Pull Request

Den4ik commented:
@sidolov @gabrieldagama Could you please check and assign priority to this PR.
I believe that we can assign High Priority and merge it because it's really improve developer experience

magento-engcom-team commented:
Hi @bgorski, thank you for the review.
[ENGCOM-9223](https://jira.corp.magento.com/browse/ENGCOM-9223) has been created to process this Pull Request

Den4ik committed:
Merge branch '2.4-develop' into add-phpmd-rule-unusedformalparameter-with-plugin-ignore

Den4ik commented:
@magento run all tests

Den4ik committed:
Merge branch '2.4-develop' into add-phpmd-rule-unusedformalparameter-with-plugin-ignore

Den4ik commented:
@magento run all tests

engcom-Echo committed:
Merge branch '2.4-develop' into add-phpmd-rule-unusedformalparameter-with-plugin-ignore

engcom-Echo commented:
@magento run all tests

engcom-Echo commented:
@magento run Unit Tests ,Functional Tests CE,Functional Tests EE,Functional Tests B2B

engcom-Echo commented:
@magento create issue

engcom-Echo commented:
### :heavy_check_mark: QA Passed

**Manual testing scenario:**

1.  Configure phpmd
2. Remove `@SuppressWarnings(PHPMD.UnusedFormalParameter)` from some plugin
3. Add `PluginInterface` implementation into the class
4. Run `phpmd`. Expected no `UnusedFormalParameter` errors


**Before:** :heavy_multiplication_x: No information of the invoice email was shown in the header.
<img width="1000" alt="Screenshot 2023-08-24 at 9 03 55 AM" src="https://github.com/magento/magento2/assets/51681435/3109edf0-148e-44b2-812e-ce2e81af2b97">

**After:** :heavy_check_mark:  Mail sent information was now getting displayed.
<img width="779" alt="Screenshot 2023-08-24 at 9 11 50 AM" src="https://github.com/magento/magento2/assets/51681435/6769c215-f23a-4021-995e-1d8feb219f89">

Builds are failed. Hence, moving this PR to **Extended Testing.**

engcom-Echo commented:
@magento run Unit Tests,Functional Tests CE,Functional Tests EE,Functional Tests B2B

engcom-Echo commented:
@magento run Unit Tests,Functional Tests EE,Functional Tests CE ,Functional Tests B2B

engcom-Echo commented:
Functional Tests B2B failure is different on last two run on same commit.
Run1
<img width="1440" alt="Screenshot 2023-08-25 at 10 14 05 AM" src="https://github.com/magento/magento2/assets/51681435/224df84e-6067-4623-b658-408e43faa6bf">

Run2
<img width="1397" alt="Screenshot 2023-08-25 at 10 14 11 AM" src="https://github.com/magento/magento2/assets/51681435/10c6d63d-85d4-49b7-9ed5-89a6f4a21331">

Hence moving it to `Merge In Progress`.
